The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12571837970/job/35043201964) [(plain)](https://github.com/rust-lang/rust/commit/dbad0c83c16eb78c59d9d20bdc2c0c8162a003e3/checks/35043201964/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.771 Building wheels for collected packages: reuse
#13 2.772   Building wheel for reuse (pyproject.toml): started
#13 3.016   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.018   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.018   Stored in directory: /tmp/pip-ephem-wheel-cache-3663gcoj/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.020 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.406 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.407 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.927 Collecting virtualenv
#13 3.927 Collecting virtualenv
#13 3.964   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.030      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 66.5 MB/s eta 0:00:00
#13 4.087 Collecting filelock<4,>=3.12.2
#13 4.090   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.124 Collecting platformdirs<5,>=3.9.1
#13 4.127   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.144 Collecting distlib<1,>=0.3.7
#13 4.156      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 76.6 MB/s eta 0:00:00
#13 4.156      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 76.6 MB/s eta 0:00:00
#13 4.236 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.420 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      114624 kB
DirectMap2M:     7225344 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
Diff in /checkout/library/std/src/sys/pal/windows/os.rs:5:
 #[cfg(test)]
 mod tests;
 
-use super::api;
 #[cfg(not(target_vendor = "uwp"))]
 use super::api::WinError;
-use super::to_u16s;
+use super::{api, to_u16s};
 use crate::ffi::{OsStr, OsString};
 use crate::os::windows::ffi::EncodeWide;
 use crate::os::windows::ffi::EncodeWide;
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/sys/exit_guard.rs" "/checkout/library/std/src/sys/pal/unsupported/pipe.rs" "/checkout/library/std/src/sys/pal/unsupported/thread.rs" "/checkout/library/std/src/sys/pal/unsupported/net.rs" "/checkout/library/std/src/sys/pal/unsupported/process.rs" "/checkout/library/std/src/sys/pal/unsupported/env.rs" "/checkout/library/std/src/sys/pal/unsupported/os.rs" "/checkout/library/std/src/sys/pal/unsupported/io.rs" "/checkout/library/std/src/sys/pal/unsupported/common.rs" "/checkout/library/std/src/sys/pal/unsupported/fs.rs" "/checkout/library/std/src/sys/pal/unsupported/mod.rs" "/checkout/library/std/src/sys/pal/unsupported/args.rs" "/checkout/library/std/src/sys/pal/unsupported/time.rs" "/checkout/library/std/src/sys/pal/unsupported/stdio.rs" "/checkout/library/std/src/sys/pal/hermit/thread.rs" "/checkout/library/std/src/sys/pal/hermit/net.rs" "/checkout/library/std/src/sys/pal/hermit/futex.rs" "/checkout/library/std/src/sys/pal/hermit/env.rs" "/checkout/library/std/src/sys/pal/hermit/fd.rs" "/checkout/library/std/src/sys/pal/hermit/os.rs" "/checkout/library/std/src/sys/pal/hermit/io.rs" "/checkout/library/std/src/sys/pal/hermit/fs.rs" "/checkout/library/std/src/sys/pal/hermit/mod.rs" "/checkout/library/std/src/sys/pal/hermit/args.rs" "/checkout/library/std/src/sys/pal/hermit/time.rs" "/checkout/library/std/src/sys/pal/hermit/stdio.rs" "/checkout/library/std/src/sys/pal/windows/pipe.rs" "/checkout/library/std/src/sys/pal/windows/thread.rs" "/checkout/library/std/src/sys/pal/windows/stack_overflow.rs" "/checkout/library/std/src/sys/pal/windows/net.rs" "/checkout/library/std/src/sys/pal/windows/stdio/tests.rs" "/checkout/library/std/src/sys/pal/windows/process.rs" "/checkout/library/std/src/sys/pal/windows/handle.rs" "/checkout/library/std/src/sys/pal/windows/futex.rs" "/checkout/library/std/src/sys/pal/windows/os/tests.rs" "/checkout/library/std/src/sys/pal/windows/c/windows_sys.rs" "/checkout/library/std/src/sys/pal/windows/args/tests.rs" "/checkout/library/std/src/sys/pal/windows/handle/tests.rs" "/checkout/library/std/src/sys/pal/windows/env.rs" "/checkout/library/std/src/sys/pal/windows/api/tests.rs" "/checkout/library/std/src/sys/pal/windows/os.rs" "/checkout/library/std/src/sys/pal/windows/io.rs" "/checkout/library/std/src/sys/pal/windows/fs.rs" "/checkout/library/std/src/sys/pal/windows/api.rs" "/checkout/library/std/src/sys/pal/windows/process/tests.rs" "/checkout/library/std/src/sys/pal/windows/stack_overflow_uwp.rs" "/checkout/library/std/src/sys/pal/windows/mod.rs" "/checkout/library/std/src/sys/pal/windows/compat.rs" "/checkout/library/std/src/sys/pal/windows/args.rs" "/checkout/library/std/src/sys/pal/windows/time.rs" "/checkout/library/std/src/sys/pal/windows/fs/remove_dir_all.rs" "/checkout/library/std/src/sys/pal/windows/c.rs" "/checkout/library/std/src/sys/pal/windows/stdio.rs" "/checkout/library/std/src/sys/pal/teeos/thread.rs" "/checkout/library/std/src/sys/pal/teeos/net.rs" "/checkout/library/std/src/sys/pal/teeos/os.rs" "/checkout/library/std/src/sys/pal/teeos/mod.rs" "/checkout/library/std/src/sys/pal/teeos/stdio.rs" "/checkout/library/std/src/sys/pal/sgx/thread.rs" "/checkout/library/std/src/sys/pal/sgx/net.rs" "/checkout/library/std/src/sys/pal/sgx/libunwind_integration.rs" "/checkout/library/std/src/sys/pal/sgx/env.rs" "/checkout/library/std/src/sys/pal/sgx/fd.rs" "/checkout/library/std/src/sys/backtrace.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Jan  1 15:16:30 UTC 2025
  network time: Wed, 01 Jan 2025 15:16:30 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/library/std/src/sys/pal/windows/os.rs
+++ b/library/std/src/sys/pal/windows/os.rs
@@ -5,8 +5,9 @@
 #[cfg(test)]
 mod tests;
 
-use super::api::{self, WinError};
-use super::to_u16s;
+#[cfg(not(target_vendor = "uwp"))]
+use super::api::WinError;
+use super::{api, to_u16s};
 use crate::error::Error as StdError;
 use crate::ffi::{OsStr, OsString};
 use crate::os::windows::ffi::EncodeWide;

--- a/src/doc/rustc/src/SUMMARY.md
+++ b/src/doc/rustc/src/SUMMARY.md
@@ -87,6 +87,7 @@
     - [*-unknown-openbsd](platform-support/openbsd.md)
     - [*-unknown-redox](platform-support/redox.md)
     - [\*-unknown-uefi](platform-support/unknown-uefi.md)
+    - [\*-uwp-windows-msvc](platform-support/uwp-windows-msvc.md)
     - [\*-wrs-vxworks](platform-support/vxworks.md)
     - [wasm32-wasip1](platform-support/wasm32-wasip1.md)
     - [wasm32-wasip1-threads](platform-support/wasm32-wasip1-threads.md)

--- a/src/doc/rustc/src/platform-support.md
+++ b/src/doc/rustc/src/platform-support.md
@@ -264,7 +264,7 @@ target | std | host | notes
 [`aarch64-unknown-redox`](platform-support/redox.md) | ✓ |  | ARM64 Redox OS
 [`aarch64-unknown-teeos`](platform-support/aarch64-unknown-teeos.md) | ? |  | ARM64 TEEOS |
 [`aarch64-unknown-trusty`](platform-support/trusty.md) | ? |  |
-`aarch64-uwp-windows-msvc` | ✓ |  |
+[`aarch64-uwp-windows-msvc`](platform-support/uwp-windows-msvc.md) | ✓ |  |
 [`aarch64-wrs-vxworks`](platform-support/vxworks.md) | ✓ |  | ARM64 VxWorks OS
 `aarch64_be-unknown-linux-gnu` | ✓ | ✓ | ARM64 Linux (big-endian)
 `aarch64_be-unknown-linux-gnu_ilp32` | ✓ | ✓ | ARM64 Linux (big-endian, ILP32 ABI)
@@ -312,7 +312,7 @@ target | std | host | notes
 [`i686-unknown-openbsd`](platform-support/openbsd.md) | ✓ | ✓ | 32-bit OpenBSD [^x86_32-floats-return-ABI]
 [`i686-unknown-redox`](platform-support/redox.md) | ✓ |  | i686 Redox OS
 `i686-uwp-windows-gnu` | ✓ |  | [^x86_32-floats-return-ABI]
-`i686-uwp-windows-msvc` | ✓ |  | [^x86_32-floats-return-ABI]
+[`i686-uwp-windows-msvc`](platform-support/uwp-windows-msvc.md) | ✓ |  | [^x86_32-floats-return-ABI]
 [`i686-win7-windows-msvc`](platform-support/win7-windows-msvc.md) | ✓ |   | 32-bit Windows 7 support [^x86_32-floats-return-ABI]
 [`i686-wrs-vxworks`](platform-support/vxworks.md) | ✓ |  | [^x86_32-floats-return-ABI]
 [`loongarch64-unknown-linux-ohos`](platform-support/openharmony.md) | ✓ |   | LoongArch64 OpenHarmony
@@ -383,8 +383,8 @@ target | std | host | notes
 [`thumbv4t-none-eabi`](platform-support/armv4t-none-eabi.md) | * |  | Thumb-mode Bare Armv4T
 [`thumbv5te-none-eabi`](platform-support/armv5te-none-eabi.md) | * |  | Thumb-mode Bare Armv5TE
 [`thumbv6m-nuttx-eabi`](platform-support/nuttx.md) | * |  | ARMv6M with NuttX
-`thumbv7a-pc-windows-msvc` | ✓ |  |
-`thumbv7a-uwp-windows-msvc` | ✓ |  |
+`thumbv7a-pc-windows-msvc` |  |  |
+[`thumbv7a-uwp-windows-msvc`](platform-support/uwp-windows-msvc.md) |  |  |
 [`thumbv7em-nuttx-eabi`](platform-support/nuttx.md) | * |  | ARMv7EM with NuttX
 [`thumbv7em-nuttx-eabihf`](platform-support/nuttx.md) | * |  | ARMv7EM with NuttX, hardfloat
 [`thumbv7m-nuttx-eabi`](platform-support/nuttx.md) | * |  | ARMv7M with NuttX
@@ -406,7 +406,7 @@ target | std | host | notes
 [`x86_64-unknown-openbsd`](platform-support/openbsd.md) | ✓ | ✓ | 64-bit OpenBSD
 [`x86_64-unknown-trusty`](platform-support/trusty.md) | ? |  |
 `x86_64-uwp-windows-gnu` | ✓ |  |
-`x86_64-uwp-windows-msvc` | ✓ |  |
+[`x86_64-uwp-windows-msvc`](platform-support/uwp-windows-msvc.md) | ✓ |  |
 [`x86_64-win7-windows-msvc`](platform-support/win7-windows-msvc.md) | ✓ |   | 64-bit Windows 7 support
 [`x86_64-wrs-vxworks`](platform-support/vxworks.md) | ✓ |  |
 [`x86_64h-apple-darwin`](platform-support/x86_64h-apple-darwin.md) | ✓ | ✓ | macOS with late-gen Intel (at least Haswell)

--- a/src/doc/rustc/src/platform-support/uwp-windows-msvc.md
+++ b/src/doc/rustc/src/platform-support/uwp-windows-msvc.md
@@ -1,0 +1,52 @@
+# `x86_64-uwp-windows-msvc`, `i686-uwp-windows-msvc`, `thumbv7a-uwp-windows-msvc` and `aarch64-uwp-windows-msvc`
+
+**Tier: 3**
+
+Windows targets for Universal Windows Platform (UWP) applications, using MSVC toolchain.
+
+## Target maintainers
+
+- [@bdbai](https://github.com/bdbai)
+
+## Requirements
+
+These targets are cross-compiled with std support. The host requirement and
+binary format are the same as the corresponding non-UWP targets (i.e.
+`x86_64-pc-windows-msvc`, `i686-pc-windows-msvc`, `thumbv7a-pc-windows-msvc`
+and `aarch64-pc-windows-msvc`).
+
+## Building the targets
+
+The targets can be built by enabling them for a `rustc` build, for example:
+
+```toml
+[build]
+build-stage = 1
+target = ["x86_64-uwp-windows-msvc", "aarch64-uwp-windows-msvc"]
+```
+
+## Building Rust programs
+
+Rust does not yet ship pre-compiled artifacts for these targets. To compile for
+these targets, you will either need to build Rust with the targets enabled (see
+"Building the targets" above), or build your own copy of `std` by using
+`build-std` or similar.
+
+Example of building a Rust project for x64 UWP using `build-std`:
+
+```pwsh
+cargo build -Z build-std=std,panic_abort --target x86_64-uwp-windows-msvc
+```
+
+## Testing
+
+Currently there is no support to run the rustc test suite for this target.
+
+## Cross-compilation toolchains and C code
+
+In general, the toolchain target should match the corresponding non-UWP
+targets. Beware that not all Win32 APIs behave the same way in UWP, and some
+are restricted in [AppContainer](https://learn.microsoft.com/en-us/windows/win32/secauthz/appcontainer-for-legacy-applications-)
+or even not available at all. If the C code being compiled happens to use any
+of restricted or unavailable APIs, consider using allowed alternatives or
+disable certain feature sets to avoid using them.

